client/cli/src/runner.rs

```diff
@@ -27,17 +27,21 @@ use log::info;
 use sc_service::{Configuration, TaskType, TaskManager};
 use sp_utils::metrics::{TOKIO_THREADS_ALIVE, TOKIO_THREADS_TOTAL};
 use std::marker::PhantomData;
+use std::error;
+use std::result;
+use sc_service::Error as ServiceError;
+use crate::error::Error as CliError;
 
 #[cfg(target_family = "unix")]
-async fn main<F, E>(func: F) -> std::result::Result<(), Box<dyn std::error::Error>>
+async fn main<F, E>(func: F) -> std::result::Result<(), E>
 where
 	F: Future<Output = std::result::Result<(), E>> + future::FusedFuture,
-	E: 'static + std::error::Error,
+	E: error::Error + Send + Sync + 'static + From<ServiceError>,
 {
 	use tokio::signal::unix::{signal, SignalKind};
 
-	let mut stream_int = signal(SignalKind::interrupt())?;
-	let mut stream_term = signal(SignalKind::terminate())?;
+	let mut stream_int = signal(SignalKind::interrupt()).map_err(ServiceError::Io)?;
+	let mut stream_term = signal(SignalKind::terminate()).map_err(ServiceError::Io)?;
```

Comment:
I think the "right" way to achieve this PR goal would be to add `ServiceError::Io` to `sc_cli::Error` so it can use `?` easily.

Then you can replace all the `Box<dyn Error>` to `sc_cli::Error`. And remove all the generic type of errors.

What do you think?

Reply:
The PR goal is to be able to use it with polkadot (which has different error types!) as such no matter what error type is introduced, it will not avoid the `Error -> String -> Error` round trip which removes the source, which is the motivation for this PR.

Reply:
Oh I see! Ok my bad

Reply:
No worries at all :)